### ISSUE TITLE
Implement Local "Loss"

### ISSUE THREAD

**Issue by gabriel-q-wang** [ezCGP/ezCGP#225]
Local loss in our case uses a combination of a global and local discriminator and adding their losses.
Technically I think this makes it closer to a "ensemble" of two discriminator models.

I think our current implementation uses one refiner, one discriminator, and one training config block. 
To implement the local discriminator I think preferably we add an optional block for the local discriminator, which can mate/mutate with other local discriminators. Or we can just have the local discriminator clone the global discriminator config.  

Might run into high computation costs.

**Comment by roddtalebi** [comment]
2 things:

1. I'm pretty sure I coded ezcgp so that one of the base assumptions is that the number of blocks is fixed and the order in which they get computed is fixed. We could try to get creative but if it is a whole third network that we may or may not want to use then it will take some big code changes to make it work
2. I'd check with Jason to see how we should prioritize a change or implementation like this

**Comment by gabriel-q-wang** [comment]
So what we could do is having the local discriminator copy the global discriminator's architecture, and just change the input window size? That way we're really only evolving one architecture, it just happens to be used twice with one minor tweak? Or is that still not possible in ezcgp?

**Comment by roddtalebi** [comment]
If there is a way to copy a pytorch graph and edit it in-place, that would be super clutch; then we can move with that idea. There could still be other options but not immediately obvious.
Here is where we extract the built graphs and where we would want to copypaste + adjust the discriminator block for the local loss
https://github.com/ezCGP/ezCGP/blob/feature/simgan-support/codes/individual_definitions/individual_evaluate.py#L385

**Comment by roddtalebi** [comment]
NOTE right now we have 3 blocks: Refiner architecture, Discriminator architecture, TrainConfig dictionary

After talked to Jason, we concluded that we have 2 options in front of us:

1. If our (non-evolution) testing shows that the discriminator and local-discriminator networks work better with the exact same architectures (outside of the shapes of the input data), then we will continue to have simgan with 3 blocks and then somehow produce a 3rd graph for the local-discriminator using the same genetic material as the discriminator block and then rely on the trainconfig dict to have a boolean flag to turn that local-discriminator graph on/off
2. else if we find that results are better when we force the local-discriminator to have a different architecture from the discriminator, then we will have 4 blocks with the newest being for the local-discriminator architecture but continue to rely on the trainconfig to have a boolean flag to decide whether we train/use that new network or not

@gabriel-q-wang  and @anishthite are volun-told do do the testing to inform how we structure our simgan genome to allow for local-discriminator network

**Comment by anishthite** [comment]
Ran experiments with 1. the same resnet structure, 2. a smaller kernel for resnet in local D, 3. transformer D and resnet local D in order to see the difference. 

There was a significant difference in runs 1/2 and 3, but not much between 1 and 2. I could maybe have been not changing the resnet enough for it to matter. 

Conclusion: Making a drastic change like changing the entire architecture will make a difference but smaller modifications of a single architecture will likely not make a difference. I'll try playing around with the other boolean flags of the resnet to try to get it to do something

**Comment by gabriel-q-wang** [comment]
@anishthite Wait how did you do runs with the local discriminator? Rodd and I were discussing implementation details earlier, and I don't think it's been implemented in the code yet unless you did it already

**Comment by anishthite** [comment]
I did it on the og simgan repo, just manually changing the config file

**Comment by anishthite** [comment]
> Ran experiments with 1. the same resnet structure, 2. a smaller kernel for resnet in local D, 3. transformer D and resnet local D in order to see the difference.
> 
> There was a significant difference in runs 1/2 and 3, but not much between 1 and 2. I could maybe have been not changing the resnet enough for it to matter.
> 
> Conclusion: Making a drastic change like changing the entire architecture will make a difference but smaller modifications of a single architecture will likely not make a difference. I'll try playing around with the other boolean flags of the resnet to try to get it to do something

Ran some more trials and reached the same conclusion: only large-scale architecture changes like resnet->transformer are enough to affect the result

**Comment by roddtalebi** [comment]
@gabriel-q-wang 
I'm gonna try to work on implementing the 'ezcgp' part of the code so that you can plug in your local loss stuff. 
I see you made this branch, right?
https://github.com/ezCGP/ezCGP/commits/origin/feature/simgan-support/Issue225-Local-Loss

**Comment by roddtalebi** [comment]
- adding the boolean flag to turn on 'local loss' was easy
- going to play with the [evaluate method for discriminator](https://github.com/ezCGP/ezCGP/blob/origin/feature/simgan-support/Issue225-Local-Loss/codes/block_definitions/evaluate/block_evaluate_pytorch.py#L310) to see if I can build the discriminator and the 'local' discriminator
- then I'll have to change how the outputs of each block are accepted at the individual level [here](https://github.com/ezCGP/ezCGP/blob/origin/feature/simgan-support/Issue225-Local-Loss/codes/individual_definitions/individual_evaluate.py#L384) ...it'll prob be something like always looking for a local discriminator network and either setting that variable equal to the graph we get from the block.output or just setting it as None, so expect to use a variable like that in the rest of the code

The rest @gabriel-q-wang, will hopefully be all you in implementing the local loss stuff around some ```if untrained_local_discriminator is not None:``` or something

Also, if the 'normal' shape of the data is (batchsize, 1, 92), what will the input shape of the data to the local network? (batchsize, 1, WINDOW)? should/can 'window' be hardcoded deep in the code somewhere or should that be more easily accessible so that we can quickly change the number to play with our networks, ie should a user be able to change that in the problem file?

**Comment by roddtalebi** [comment]
pushed up my changes...
https://github.com/ezCGP/ezCGP/commit/f71909fb45e4eed436a658a7d84ae8b9e6f747be

Big question is how/what are we going to pass in [here](https://github.com/ezCGP/ezCGP/blob/origin/feature/simgan-support/Issue225-Local-Loss/codes/block_definitions/evaluate/block_evaluate_pytorch.py#L311) to be the input of the local loss network.

**Comment by roddtalebi** [comment]
@gabriel-q-wang 
https://github.com/ezCGP/ezCGP/tree/origin/feature/simgan-support/Issue225-Local-Loss ...right branch?

[Let me get that code](https://www.youtube.com/watch?v=vCadcBR95oU)

**Comment by roddtalebi** [comment]
There are layers/primitives hard-coded in to the end of both the [refiner](https://github.com/ezCGP/ezCGP/blob/origin/feature/simgan-support/Issue225-Local-Loss/codes/block_definitions/evaluate/block_evaluate_pytorch.py#L236) and [discriminator](https://github.com/ezCGP/ezCGP/blob/origin/feature/simgan-support/Issue225-Local-Loss/codes/block_definitions/evaluate/block_evaluate_pytorch.py#L282) blocks

**Comment by gabriel-q-wang** [comment]
Pushed up the code, getting the error on the following line: https://github.com/ezCGP/ezCGP/blob/origin/feature/simgan-support/Issue225-Local-Loss/codes/individual_definitions/individual_evaluate.py#L525 . Going to double check the network by itself, and maybe change how the input was handled since it was kinda gimmicky. From my understanding the graph is build using the default method and the shape of the input layer is defined by the shape of the input data. I didn't actually change the input data but just forced the shape to be different, which I thought would make the network properly, but I might be wrong there. If so I might need to make a copy of the data, splice it to properly reflect the local input, and then call the build_graph method. https://github.com/ezCGP/ezCGP/blob/565bc9e7cf938b2155a1efa6ef4ec2a55038b309/codes/block_definitions/evaluate/block_evaluate_pytorch.py#L315

Also noted a few other errors while writing/testing:

1) I think some of the ArgumentTypes are broken when trying to seed from a lisp. Added a new section size parameter, but only ArgumentType_Int1to5 seemed to work, switching it to ArgumentType_Ints or a different range of ints lead to errors. https://github.com/ezCGP/ezCGP/blob/565bc9e7cf938b2155a1efa6ef4ec2a55038b309/codes/block_definitions/utilities/operators_simgan_train_config.py#L88 

2) If all the individuals are dead/invalid, there is an error when trying to add to the hall of fame, as there are no individuals to add, which causes the whole process to fail.

**Comment by roddtalebi** [comment]
RE the other errors:

1. This is because we have to first make those argtypes available to the individual's genome via the [BlockArgument class](https://github.com/ezCGP/ezCGP/blob/565bc9e7cf938b2155a1efa6ef4ec2a55038b309/codes/block_definitions/operators/block_operators.py#L340). Funny enough though, I was lying in bed the other day thinking that I should just rewrite the BlockArgument class or make a new one that autofills the block with the arguments required by the operators listed in BlockOperators; then I forgot about this until now...went ahead and made [an issue](https://github.com/ezCGP/ezCGP/issues/253) to eventually get to it 
2. made it into [an issue](https://github.com/ezCGP/ezCGP/issues/252) to catch this edge case. nice catch.

Looking at the logs, it looks like it errors on feature_extraction(), but it runs successfully several times before it finally errors which is weird. My initial instinct is that it has something to do with the number of features being so small (4). I'm going to hardcode it to something like 100 and see what happens. Also I'm going to change the way we set the 'shape' of the input layer before we build the local loss network, to something that is more pleasing to my mind.

**Comment by roddtalebi** [comment]
@gabriel-q-wang , I got it!
https://github.com/ezCGP/ezCGP/blob/565bc9e7cf938b2155a1efa6ef4ec2a55038b309/misc/features.py#L119

If we are clipping our data to something smaller than 40 features, then it'll try to find the peak in data[:, 40:80] which will just return an empty array and an argmax of an empty array is going to error. What does your feature extraction code look like? Or do you guys always clip the data to something bigger than 40? I assume feature extraction method has to be different or else how would the method be useful?

**Comment by gabriel-q-wang** [comment]
OHHHHH I remember now. I ran into this issue when I was merging with the other repo as well. I disabled the feature extraction for all local discriminators, like you said it doesn't make sense to look for the same features on a smaller segment. So it's not an exact clone of the discriminator, since feature extraction needs to be disabled. I'll think about how to implement that...

**Comment by roddtalebi** [comment]
hmm...that'll be tough to do for any new network that gets evolved...couple options i could thing of at the top of my head...

1. could take the lazy approach...in feature_extraction method, if the size of the data is < 92 then just pass and do nothing. only problem with that is that it will try to concat the output of feature_extraction with output of minibatch_discriminator and all the other convolutions but maybe thats okay
2. could try to deepcopy blockmaterial [here before we build the graph](https://github.com/ezCGP/ezCGP/blob/565bc9e7cf938b2155a1efa6ef4ec2a55038b309/codes/block_definitions/evaluate/block_evaluate_pytorch.py#L316) and edit the genome by replacing any node with feature_extraction as a method with another method that does nothing or returns empty array which would solve the concat issue from other step

**Comment by roddtalebi** [comment]
@gabriel-q-wang fyi checkout the changes I made to the individual_evaluate.py script. subtle but important in making sure your code works for when the program eventually decides to mutate 'local_section_size' but not 'train_local_loss'.
https://github.com/ezCGP/ezCGP/commit/95930a6e1d79796739d2976daa0ddc677c567bc6

my changes to codes/block_definitions/evaluate/block_evaluate_pytorch.py are totally bogus and what you had was exactly the same. I just had an idea to make fake data with np but I'd rather the shape have 'None' in it for batch_size than some bogus number I picked just to make fake data and I won't be able to do that with numpy so I tried going back to your original code.

**Comment by gabriel-q-wang** [comment]
The four objectives I'd recommend running on are: fid, ks_stat, num_sig, and avg_feat_pval

**Comment by roddtalebi** [comment]
@gabriel-q-wang, did I get the min/max right for each of those?
fid, ks_stat, num_sig -> min
avg_feat_pval -> max

**Comment by gabriel-q-wang** [comment]
Yep those are all right. Also oops realized this was the wrong issue.

**Comment by gabriel-q-wang** [comment]
Comment from the PR: My implementation might be a little hacky. The main thing is that since the local discriminator shouldn't use the feature extraction, the resultant shape from MDB+Convs+Features is not equal to what the linear layer expects. So I made a "dynamic" linear layer who's size depends on the incoming input. But this is weird since we are adding a new layer to the network during the first forward pass, and the layer needs to be on the same device as the model, but it's kinda hard to pass that as a parameter. So for now I'm assuming that we'll always be using a GPU and the device is always cuda.

**Comment by roddtalebi** [comment]
I don't like it.

What if feature_extraction just returns a tensor of all 0s? shape will be maintained but, at least in my head (the head that hasn't thought about details of backprop + nonlinear activations in a long time), it shouldn't affect the overall process too much

**Comment by roddtalebi** [comment]
Having a real moment here..."Maybe we need to change the way we are approaching the problem"

Can we make the assumption that feature_extraction will only ever be useful when the input to the primitive is the original input data to the discriminator block?
If so, then we shouldn't let it be a primitive in the architecture that can moved around and can have just about anything added to it. Let's just make it into another block like train_config, where we mutate the args to toggle what get's turned on and off in the feature_extraction, and then the discriminator block will have 2 inputs: original data, extracted features.

**Comment by gabriel-q-wang** [comment]
Hmmm I think the main problem is how we/pytorch handles concatenation. Since in our "default" architecture we concatenate the outputs from the convolutions, minibatch discrimination, and feature extraction. In the other simgan repo we could toggle all those options if we didn't want to include them, but since pytorch requires knowing what the input shape would be we would end up changing the expected output shape for the next linear layer manually for those options.

For ezCGP since we want to do neural architecture search, I think we'd still run into a lot of shape mismatches throughout the process, and the solution would either be to force an expected shape or some dynamic layer to handle it. My main issue with forcing the expected shape is that I don't know if we would constrain the search space too much. If any changes occur that would change the shape of the output, like different convolutional parameters, another concatenate layer, etc. I think it would end up breaking. So I think that the most drastic changes that could occur during evolution would be fairly small hyperparameter changes and not too much in terms of architecture.

I think your idea of having the extracted features as a separate input is a good idea, as the feature extraction should only be run on the original data. But I don't think it really fixes the underlying problem.

**Comment by roddtalebi** [comment]
Changing feature extraction to return a tensor of 0s with the right shape instead of an empty tensor SHOULD* solve everything...

*it is too simple, there must be a problem, right?!

**Comment by gabriel-q-wang** [comment]
Returning the tensors of 0s is the easiest solution, but I'm not sure how it'd affect things. I could run a few examples and see what happens. But at least to me returning a vector of all zeros (which could be valid values for the features) isn't the same as not using the feature extraction. But right now the first linear layer always expects an input of size 42, (32 from convs + 10 from feature extraction). Sure we could always return the right size for the feature extraction but can we always guarantee the input size of 32 is coming from the convolutional architecture? 


Thoughts for the future but could we split the build graph portion into two parts? One to handle convolutions and concatenates, another for the final linear classifier portions? We kinda already do that, but it's all done in one step, so if the shape mismatches it throws an error. But could it build the conv stuff first, get the expected output shape, then append the linear layers to fit that output in a second build graph step?

**Comment by roddtalebi** [comment]
I can get the sizes working right, I'm sure of it.

Also I did this and it convinced me that it'll work...

Output from convs -> (batch_size, 32)
Output from feature extract -> (batch_size, 10)
Output from concat -> (batch_size, 42)
Linear layer shape -> (42, 20)
Output of linear linear -> (batch_size, 20)

If feature extraction is all 0s, then the last 10 columns of 'output from concat' will all be 0s. The next step is just a super linear layer, ie just matrix multiplication.
So if feature extraction returns all 0s, the last 10 rows of the linear layer won't affect the matrix multiplication at all since they will always be multiplied by 0.
So doing matrix (batchsize, 42) x (42,20) with the first matrix having last 10 columns all 0, is the same as doing (batchsize,32) x *(32,20)

**Comment by roddtalebi** [comment]
https://github.com/ezCGP/ezCGP/commits/feature/simgan_support/Issue225-Try_return_0s
